h2/src/main/org/h2/mvstore/tx/TransactionStore.java

```diff
@@ -50,7 +51,9 @@
      * <p>
      * Key: opId, value: [ mapId, key, oldValue ].
      */
-    final MVMap<Long, Object[]> undoLog;
+    @SuppressWarnings("unchecked")
+    final MVMap<Long,Object[]> undoLogs[] = new MVMap[MAX_OPEN_TRANSACTIONS];
```

Comment:
It's better to declare it as `MVMap<Long,Object[]>[] undoLogs`.

I noticed that unused elements of this array are not cleared. If a lot of sessions were opened and closed and application does not use so many sessions at once later this creates a memory leak.

Reply:
If element not in use all what it holds is a reference to an empty MVMap, which is probably 200-300 bytes at most. I do not want to reallocate maps for frequently used array elements (with low indexes). High-positioned (infrequently used) elements can be either closed and nullified right away or periodically  swept. This also involves closing the map and related synchronization. Need to come up with some heuristic what is worth keeping, and what's not.

Reply:
OK, a few megabytes for an application that allocated a 50,000 sessions at once is not a big price.

I didn't check the retained sizes of these maps before, they share some fields, so total size is not so large.

Reply:
They should share a single empty root page